<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Taxonomic Assignment</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">dada2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="dada-installation.html">Install</a>
</li>
<li>
  <a href="tutorial.html">Tutorial</a>
</li>
<li>
  <a href="bigdata.html">Big Data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://www.bioconductor.org/packages/3.5/bioc/manuals/dada2/man/dada2.pdf">Manual</a>
    </li>
    <li>
      <a href="assign.html">Taxonomy</a>
    </li>
    <li>
      <a href="pool.html">Pooling</a>
    </li>
    <li>
      <a href="faq.html">FAQ</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Evaluation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://dx.doi.org/10.1038/nmeth.3869">Manuscript</a>
    </li>
    <li>
      <a href="SMBS_DADA2.pdf">Symposium Slides</a>
    </li>
    <li>
      <a href="SotA.html">Benchmarking</a>
    </li>
    <li>
      <a href="https://purl.stanford.edu/mh194vj6733">Manuscript code</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.twitter.com/bejcal">
    <span class="ion ion-social-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2/issues">
    <span class="fa fa-question fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Taxonomic Assignment</h1>

</div>


<hr />
<p>The primary goal of the dada2 package is to infer the exact sequence variants in amplicon sequenced samples and their associated abundances. However, a common companion step, especially when sequencing taxonomic markers such as 16S or 18S rRNA gene or the ITS region in fungi, is to assign taxonomy to the inferred sequence variants (or OTUs) by comparing them to a reference database of sequences with known taxonomy.</p>
<p>The dada2 package provides two taxonomic assignment methods.</p>
<p><code>assignTaxonomy(...)</code> implements the RDP naive Bayesian classifier method described in <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1950982/">Wang et al. 2007</a>. In short, the kmer profile of the sequences to be classified are compared against the kmer profiles of all sequences in a training set of sequences with assigned taxonomies. The reference sequence with the most similar profile is used to assign taxonomy to the query sequence, and then a bootstrapping approach is used to assess the confidence assignment at each taxonomic level.</p>
<p>If assignment down to the species level is desired, <code>assignSpecies(...)</code> uses exact string matching against a reference database to assign <em>Genus species</em> binomials. In short, query sequence are compared against all reference sequences that had binomial genus-species nomenclature assigned, and the genus-species of all exact matches are recorded and returned if sufficiently unambiguous.</p>
<p>These two functions can be combined to produce taxonomic assignments from Kingdom down to Species with the <code>addSpecies(...)</code> convenience function. Consistent <a href="training.html">training fastas for this purpose are provided for the Silva and RDP databases</a>.</p>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<pre class="r"><code>library(dada2); packageVersion(&quot;dada2&quot;)</code></pre>
<pre><code>## [1] &#39;1.4.0&#39;</code></pre>
<p>Both <code>assignTaxonomy</code> and <code>assignSpecies</code> depend on training datasets of taxonomically assigned sequences. Appropriately <a href="training.html">formatted datasets for several popular reference databases</a> are available.</p>
<p>In order to follow along with the examples below, <a href="training.html">download</a> the <code>rdp_train_set_16.fa.gz</code> and <code>rdp_species_assignment_16.fa.gz</code> files to an appropriate location on your local machine.</p>
<p> </p>
</div>
<div id="taxonomic-assignment" class="section level2">
<h2>Taxonomic assignment</h2>
<p>Here we classify a handful of 16S sequences from the v4 region against the RDP training set. Our sequences of interest:</p>
<pre class="r"><code>seqs &lt;- c(
&quot;ACGGAGGGTCCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTTGTTAAGCGAGATGTGAAAGCCCCGGGCTCAACCTGGGAATTGCATTTCGAACTGGCGAACTAGAGTCTTGTAGAGGGGGGTAGAATTCCAGGTGTAGCGGTGAAATGCGTAGAGATCTGGAGGA&quot;,
&quot;ACAGAGGTCTCAAGCGTTGTTCGGAATCACTGGGCGTAAAGCGTGCGTAGGCTGTTTCGTAAGTCGTGTGTGAAAGGCGCGGGCTCAACCCGCGGACGGCACATGATACTGCGAGACTAGAGTAATGGAGGGGGAACCGGAATTCTCGGTGTAGCAGTGAAATGCGTAGATATCGAGAGGA&quot;,
&quot;ACGGAGGGTGCAAGCGTTAATCGGAATCACTGGGCGTAAAGCGCACGTAGGCTGTTATGTAAGTCAGGGGTGAAAGCCCACGGCTCAACCGTGGAACTGCCCTTGATACTGCACGACTCGAATCCGGGAGAGGGTGGCGGAATTCCAGGTGTAGGAGTGAAATCCGTAGATATCTGGAGGA&quot;,
&quot;ACGTAGGTCCCGAACGTTGCGCGAATTTACTGGGCGTAAAGGGTCCGTAGGCGGTTTAGCAAGTGGTTGGTGAAATTTCACGGCTCAACCGTGAAACTGCCTTCCAAACTGCTAAACTTGAGGCAGGGAGAGGTCGGCGGAATTCCCGGTGTAGCGGTGAAATGCGTAGATATCGGGAGGA&quot;,
&quot;ACGGAGGGGGTTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGTACGTAGGCGGATTGGAAAGTTGGGGGTGAAATCCCAGGGCTCAACCCTGGAACTGCCTCCAAAACTATCAGTCTAGAGTTCGAGAGAGGTGAGTGGAATTCCAAGTGTAGAGGTGAAATTCGTAGATATTTGGAGGA&quot;,
&quot;ACGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGTGGACAGTTAAGTCAGTTGTGAAAGTTTGCGGCTCAACCGTAAAATTGCAGTTGATACTGGCTGTCTTGAGTACAGTAGAGGTGGGCGGAATTCGTGGTGTAGCGGTGAAATGCTTAGATATCACGAAGA&quot;,
&quot;ACGTAGGGCGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGAGCTCGTAGGCGGCTTGTCGCGTCGACTGTGAAAACCCGTGGCTCAACTGCGGGCTTGCAGTCGATACGGGCAGGCTAGAGTTCGGTAGGGGAGACTGGAATTCCTGGTGTAGCGGTGAAATGCGCAGATATCAGGAGGA&quot;,
&quot;ACGGAGGGTGCAAGCGTTAATCGGAATGACTGGGCGTAAAGCGCACGCAGGCGGTCTGTTAAGTTGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATTCAAAACTGACAGGCTAGAGTCTCGTAGAGGGGGGTAGAATTCCAGGTGTAGCGGTGAAATGCGTAGAGATCTGGAGGA&quot;,
&quot;ACCGACGGCCCGAGTGGTGGCCACTCTTATTGGGCCTAAAGCGTCCGTAGCCGGTCCAGTAAGTCCCTGTTTAAATCCTGCGGCTTAACCGCAGGACTGGCAGGGATACTGCTGGACTTGGGACCGGGAGAGGACAAGGGTACTTCAGGGGTAGCGGTGAAATGTGTTGATCCTTGAAGGA&quot;,
&quot;ACGTAGGTCCCGAACGTTGCGCGAAATTACTGGGCGTAAAGGGTCCGTAGGCGGTCTGGTAAGTGGAAGGTGAAAGCCCACGGCTCAACCGTGGAATTGCCTTCCAAACTGCTGGACTTGAGGGCGGAAGAGGTCGGCGGAATTCCCGGTGTAGCGGTGAAATGCGTAGATATCGGGAGGA&quot;)</code></pre>
<p>We run the <code>assignTaxonomy(...)</code> function with the RDP as the reference dataset. <em>To run the following on your local machine, modify the filename in the second argument to point to where the <code>rdp_train_set_16.fa.gz</code> file sits on your computer.</em></p>
<pre class="r"><code>set.seed(100) # Initialize random number generator for reproducibility
taxa &lt;- assignTaxonomy(seqs, &quot;Training/rdp_train_set_16.fa.gz&quot;, multithread=FALSE)
unname(taxa)</code></pre>
<pre><code>##       [,1]       [,2]              [,3]                 
##  [1,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Gammaproteobacteria&quot;
##  [2,] &quot;Bacteria&quot; &quot;Verrucomicrobia&quot; &quot;Verrucomicrobiae&quot;   
##  [3,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Deltaproteobacteria&quot;
##  [4,] &quot;Bacteria&quot; &quot;Aquificae&quot;       &quot;Aquificae&quot;          
##  [5,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Alphaproteobacteria&quot;
##  [6,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot;   &quot;Bacteroidia&quot;        
##  [7,] &quot;Bacteria&quot; &quot;Actinobacteria&quot;  &quot;Actinobacteria&quot;     
##  [8,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Gammaproteobacteria&quot;
##  [9,] &quot;Archaea&quot;  &quot;Euryarchaeota&quot;   &quot;Methanococci&quot;       
## [10,] &quot;Bacteria&quot; &quot;Aquificae&quot;       &quot;Aquificae&quot;          
##       [,4]                 [,5]                  [,6]                  
##  [1,] &quot;Alteromonadales&quot;    &quot;Shewanellaceae&quot;      &quot;Shewanella&quot;          
##  [2,] &quot;Verrucomicrobiales&quot; &quot;Verrucomicrobiaceae&quot; &quot;Akkermansia&quot;         
##  [3,] &quot;Desulfovibrionales&quot; &quot;Desulfovibrionaceae&quot; &quot;Desulfovibrio&quot;       
##  [4,] &quot;Aquificales&quot;        &quot;Hydrogenothermaceae&quot; &quot;Sulfurihydrogenibium&quot;
##  [5,] &quot;Rhodobacterales&quot;    &quot;Rhodobacteraceae&quot;    &quot;Loktanella&quot;          
##  [6,] &quot;Bacteroidales&quot;      &quot;Bacteroidaceae&quot;      &quot;Bacteroides&quot;         
##  [7,] &quot;Actinomycetales&quot;    &quot;Micromonosporaceae&quot;  &quot;Salinispora&quot;         
##  [8,] &quot;Enterobacteriales&quot;  &quot;Enterobacteriaceae&quot;  &quot;Dickeya&quot;             
##  [9,] &quot;Methanococcales&quot;    &quot;Methanococcaceae&quot;    &quot;Methanococcus&quot;       
## [10,] &quot;Aquificales&quot;        &quot;Hydrogenothermaceae&quot; &quot;Persephonella&quot;</code></pre>
<p>The return value of <code>assignTaxonomy(...)</code> is a character matrix, with each row corresponding to an input sequence, and each column corresponding to a taxonomic level. For the RDP training set, the levels are Kingdom, Phylum, Class, Order, Family, Genus, and we see that each sequence has been classified all the way to genus.</p>
<p>An important parameter to consider when running <code>assignTaxonomy(...)</code> is <code>minBoot</code>, which sets the minimum bootstrapping support required to return a taxonomic classification. The original paper recommended a threshold of 50 for sequences of 250nts or less (as these are) but a threshold of 80 generally. Let’s see what happens when we classify at that higher threshold:</p>
<pre class="r"><code>set.seed(100) # Initialize random number generator for reproducibility
taxa.80 &lt;- assignTaxonomy(seqs, &quot;Training/rdp_train_set_16.fa.gz&quot;, minBoot=80)
unname(taxa.80)</code></pre>
<pre><code>##       [,1]       [,2]              [,3]                 
##  [1,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Gammaproteobacteria&quot;
##  [2,] &quot;Bacteria&quot; &quot;Verrucomicrobia&quot; &quot;Verrucomicrobiae&quot;   
##  [3,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Deltaproteobacteria&quot;
##  [4,] &quot;Bacteria&quot; &quot;Aquificae&quot;       &quot;Aquificae&quot;          
##  [5,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Alphaproteobacteria&quot;
##  [6,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot;   &quot;Bacteroidia&quot;        
##  [7,] &quot;Bacteria&quot; &quot;Actinobacteria&quot;  &quot;Actinobacteria&quot;     
##  [8,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Gammaproteobacteria&quot;
##  [9,] &quot;Archaea&quot;  &quot;Euryarchaeota&quot;   &quot;Methanococci&quot;       
## [10,] &quot;Bacteria&quot; &quot;Aquificae&quot;       &quot;Aquificae&quot;          
##       [,4]                 [,5]                  [,6]                  
##  [1,] &quot;Alteromonadales&quot;    &quot;Shewanellaceae&quot;      &quot;Shewanella&quot;          
##  [2,] &quot;Verrucomicrobiales&quot; &quot;Verrucomicrobiaceae&quot; &quot;Akkermansia&quot;         
##  [3,] &quot;Desulfovibrionales&quot; &quot;Desulfovibrionaceae&quot; &quot;Desulfovibrio&quot;       
##  [4,] &quot;Aquificales&quot;        &quot;Hydrogenothermaceae&quot; &quot;Sulfurihydrogenibium&quot;
##  [5,] &quot;Rhodobacterales&quot;    &quot;Rhodobacteraceae&quot;    NA                    
##  [6,] &quot;Bacteroidales&quot;      &quot;Bacteroidaceae&quot;      &quot;Bacteroides&quot;         
##  [7,] &quot;Actinomycetales&quot;    &quot;Micromonosporaceae&quot;  NA                    
##  [8,] &quot;Enterobacteriales&quot;  &quot;Enterobacteriaceae&quot;  &quot;Dickeya&quot;             
##  [9,] &quot;Methanococcales&quot;    &quot;Methanococcaceae&quot;    &quot;Methanococcus&quot;       
## [10,] &quot;Aquificales&quot;        &quot;Hydrogenothermaceae&quot; &quot;Persephonella&quot;</code></pre>
<p>At this more stringent threshold, sequence 5 and 7 are not classified at the genus level. Instead an NA is returned, indicating that less than <code>minBoot=80</code> of the 100 bootstraps returned the same genus for those sequences.</p>
</div>
<div id="species-assignment" class="section level2">
<h2>Species assignment</h2>
<p>We now run the <code>assignSpecies</code> function against the RDP species-level training set. <em>To run the following on your local machine, modify the filename in the second argument to point to where the <code>rdp_species_assignment_14.fa.gz</code> file sits on your computer.</em></p>
<p><strong><span style="color:red">As of version 1.5.2, assignSpecies has been reimplemented to be much faster – more than 100x faster for common use cases.</span></strong></p>
<pre class="r"><code>genus.species &lt;- assignSpecies(seqs, &quot;Training/rdp_species_assignment_16.fa.gz&quot;)
unname(genus.species)</code></pre>
<pre><code>##       [,1]                   [,2]            
##  [1,] &quot;Shewanella&quot;           NA              
##  [2,] &quot;Akkermansia&quot;          &quot;muciniphila&quot;   
##  [3,] &quot;Desulfovibrio&quot;        &quot;piger&quot;         
##  [4,] &quot;Sulfurihydrogenibium&quot; &quot;yellowstonense&quot;
##  [5,] &quot;Sulfitobacter&quot;        &quot;pontiacus&quot;     
##  [6,] &quot;Bacteroides&quot;          NA              
##  [7,] &quot;Salinispora&quot;          NA              
##  [8,] NA                     NA              
##  [9,] NA                     NA              
## [10,] &quot;Persephonella&quot;        &quot;marina&quot;</code></pre>
<p>By default the <code>assignSpecies</code> method only returns species assignments if there is no ambiguity, i.e. all exact matches were to the same species. However, given that we are generally working with fragments of the 16S gene, it is common that exact matches are made to multiple sequences that are identical over the sequenced region. This is often still useful information, so to have all sequence hits returned the <code>returnMultiple=TRUE</code> argument can be passed to the <code>assignSpecies</code> function:</p>
<pre class="r"><code>unname(assignSpecies(seqs, &quot;Training/rdp_species_assignment_16.fa.gz&quot;, allowMultiple=TRUE))</code></pre>
<pre><code>##       [,1]                  
##  [1,] &quot;Shewanella&quot;          
##  [2,] &quot;Akkermansia&quot;         
##  [3,] &quot;Desulfovibrio&quot;       
##  [4,] &quot;Sulfurihydrogenibium&quot;
##  [5,] &quot;Sulfitobacter&quot;       
##  [6,] &quot;Bacteroides&quot;         
##  [7,] &quot;Salinispora&quot;         
##  [8,] NA                    
##  [9,] NA                    
## [10,] &quot;Persephonella&quot;       
##       [,2]                                                                                   
##  [1,] &quot;arctica/baltica/hafniensis/massilia/putrefaciens/vesiculosa&quot;                          
##  [2,] &quot;muciniphila&quot;                                                                          
##  [3,] &quot;piger&quot;                                                                                
##  [4,] &quot;yellowstonense&quot;                                                                       
##  [5,] &quot;pontiacus&quot;                                                                            
##  [6,] &quot;faecichinchillae/faecis/thetaiotaomicron&quot;                                             
##  [7,] &quot;arenicola/pacifica/tropica&quot;                                                           
##  [8,] &quot;aquatica/chrysanthemi/dadantii/dianthicola/nigrifluens/paradisiaca/roseae/solani/zeae&quot;
##  [9,] NA                                                                                     
## [10,] &quot;marina&quot;</code></pre>
<p>Several of these sequences exactly match multiple species. Setting <code>allowMultiple</code> to an integer value allows fine-grained control over this behavior. For example, <code>allowMultiple=3</code> returns up to 3 different matched species, but if 4 or more are matched it returns <code>NA</code>.</p>
</div>
<div id="kingom-to-species" class="section level2">
<h2>Kingom to Species</h2>
<p>In many cases it is desirable to link the species annotation in with the broader taxonomic annotation afforded by <code>assignTaxonomy</code>. The convenience function <code>addSpecies</code> serves this purpose, as it takes as input a taxonomy table, and outputs a table with an added species column. Only those genus-species binomials which are consistent with the genus assigned in the provided taxonomy table are retained in the output. <em>Note that we have switched to the silva training fastas here for demonstration purposes.</em></p>
<pre class="r"><code>tt &lt;- assignTaxonomy(seqs, &quot;Training/silva_nr_v128_train_set.fa.gz&quot;)
tt.plus &lt;- addSpecies(tt, &quot;Training/silva_species_assignment_v128.fa.gz&quot;, verbose=TRUE)</code></pre>
<pre><code>## 5 out of 10 were assigned to the species level.
## Of which 4 had genera consistent with the input table.</code></pre>
<pre class="r"><code>unname(tt.plus)</code></pre>
<pre><code>##       [,1]       [,2]              [,3]                 
##  [1,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Gammaproteobacteria&quot;
##  [2,] &quot;Bacteria&quot; &quot;Verrucomicrobia&quot; &quot;Verrucomicrobiae&quot;   
##  [3,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Deltaproteobacteria&quot;
##  [4,] &quot;Bacteria&quot; &quot;Aquificae&quot;       &quot;Aquificae&quot;          
##  [5,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Alphaproteobacteria&quot;
##  [6,] &quot;Bacteria&quot; &quot;Bacteroidetes&quot;   &quot;Bacteroidia&quot;        
##  [7,] &quot;Bacteria&quot; &quot;Actinobacteria&quot;  &quot;Actinobacteria&quot;     
##  [8,] &quot;Bacteria&quot; &quot;Proteobacteria&quot;  &quot;Gammaproteobacteria&quot;
##  [9,] &quot;Archaea&quot;  &quot;Euryarchaeota&quot;   &quot;Methanococci&quot;       
## [10,] &quot;Bacteria&quot; &quot;Aquificae&quot;       &quot;Aquificae&quot;          
##       [,4]                 [,5]                  [,6]                  
##  [1,] &quot;Alteromonadales&quot;    &quot;Shewanellaceae&quot;      &quot;Shewanella&quot;          
##  [2,] &quot;Verrucomicrobiales&quot; &quot;Verrucomicrobiaceae&quot; &quot;Akkermansia&quot;         
##  [3,] &quot;Desulfovibrionales&quot; &quot;Desulfovibrionaceae&quot; &quot;Desulfovibrio&quot;       
##  [4,] &quot;Aquificales&quot;        &quot;Hydrogenothermaceae&quot; &quot;Sulfurihydrogenibium&quot;
##  [5,] &quot;Rhodobacterales&quot;    &quot;Rhodobacteraceae&quot;    NA                    
##  [6,] &quot;Bacteroidales&quot;      &quot;Bacteroidaceae&quot;      &quot;Bacteroides&quot;         
##  [7,] &quot;Micromonosporales&quot;  &quot;Micromonosporaceae&quot;  &quot;Salinispora&quot;         
##  [8,] &quot;Enterobacteriales&quot;  &quot;Enterobacteriaceae&quot;  &quot;Dickeya&quot;             
##  [9,] &quot;Methanococcales&quot;    &quot;Methanococcaceae&quot;    &quot;Methanococcus&quot;       
## [10,] &quot;Aquificales&quot;        &quot;Hydrogenothermaceae&quot; &quot;Persephonella&quot;       
##       [,7]            
##  [1,] NA              
##  [2,] &quot;muciniphila&quot;   
##  [3,] &quot;piger&quot;         
##  [4,] &quot;yellowstonense&quot;
##  [5,] NA              
##  [6,] NA              
##  [7,] NA              
##  [8,] NA              
##  [9,] NA              
## [10,] &quot;marina&quot;</code></pre>
<p>For human microbiome data and V4 sequencing, we typically see a bit under half of the abundant ribosomal sequence variants (RSVs) assigned unambiguously to species-level, and a bit under 2/3rds assigned ambiguously (ie. allowing multiple species assignment). This fraction drops off in less sampled environments and for rare RSVs.</p>
</div>

<div class="container" style="width: 100%;color:grey;text-align:right">
  <hr>
  Maintained by Benjamin Callahan (benjamin DOT j DOT callahan AT gmail DOT com)
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
