<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Taxonomic reference data</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">dada2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="dada-installation.html">Install</a>
</li>
<li>
  <a href="tutorial.html">Tutorial</a>
</li>
<li>
  <a href="bigdata.html">Big Data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://www.bioconductor.org/packages/release/bioc/manuals/dada2/man/dada2.pdf">Manual</a>
    </li>
    <li>
      <a href="assign.html">Taxonomy</a>
    </li>
    <li>
      <a href="training.html">Taxonomic References</a>
    </li>
    <li>
      <a href="pool.html">Pooling</a>
    </li>
    <li>
      <a href="faq.html">FAQ</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    Evaluation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://dx.doi.org/10.1038/nmeth.3869">DADA2 Manuscript</a>
    </li>
    <li>
      <a href="http://dx.doi.org/10.1038/ismej.2017.119">Exact Sequence Variants</a>
    </li>
    <li>
      <a href="https://doi.org/10.1101/392332">PacBio Long Reads</a>
    </li>
    <li>
      <a href="SMBS_DADA2.pdf">Symposium Slides</a>
    </li>
    <li>
      <a href="SotA.html">Benchmarking</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.twitter.com/bejcal">
    <span class="ion ion-social-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2/issues">
    <span class="fa fa-question fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/benjjneb/dada2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Taxonomic reference data</h1>

</div>


<p>The <code>assignTaxonomy</code> and <code>assignSpecies</code>
functions require appropriately formatted fasta files describing the set
of taxonomically assigned sequences to use as a training dataset. We
provide appropriately formatted versions of several popular taxonomic
databases, and describe the dada2-specific format for those who wish to
use a custom database.</p>
<div id="dada2-formatted-reference-databases" class="section level2">
<h2>DADA2-formatted reference databases</h2>
<p>We maintain reference fastas for the three most common 16S databases:
Silva, RDP and GreenGenes2. The dada2 package recognizes and parses the
General Fasta releases of the UNITE project for ITS taxonomic
assignment. Formatted versions of other databases can be “contributed”
and will be made available through this page if referencable by doi (eg.
deposited at Zenodo or Figshare).</p>
<p><em>Please note that the files provided here are just reformattings
of these taxonomic databases. If using these files for taxonomic
assignment, the source database should also be cited.</em></p>
<hr />
<p><strong>Maintained:</strong></p>
<ul>
<li><p><strong><a href="https://zenodo.org/records/14169026">Silva
version 138.2</a></strong> (<a
href="https://doi.org/10.5281/zenodo.4587954">version 138.1</a>, <a
href="https://doi.org/10.5281/zenodo.1172782">version 132</a>, <a
href="https://doi.org/10.5281/zenodo.801831">version 128</a>, <a
href="http://dx.doi.org/10.5281/zenodo.158958">version 123</a>)</p>
<ul>
<li>NOTE: As of Silva version 138 the official DADA2-formatted reference
fastas are optimized for classification of Bacteria and Archaea, and are
not suitable for classifying Eukaryotes.</li>
</ul></li>
<li><p><strong><a href="https://zenodo.org/records/14168771">RDP
Trainset 19</a></strong> (<a
href="https://doi.org/10.5281/zenodo.4310150">trainset 18</a>, <a
href="https://doi.org/10.5281/zenodo.801827">trainset 16</a>, <a
href="http://dx.doi.org/10.5281/zenodo.158955">trainset 14</a>)</p></li>
<li><p><strong><a href="https://zenodo.org/records/14169078">GreenGenes2
release 2024.09</a></strong> (<a
href="http://dx.doi.org/10.5281/zenodo.158955">version 13.8 -
outdated</a>)</p></li>
<li><p><strong><a href="https://unite.ut.ee/repository.php">UNITE (use
the General Fasta releases, “All eukaryotes”)</a></strong></p></li>
</ul>
<p><strong>Contributed</strong>:</p>
<ul>
<li><p><strong><a href="https://zenodo.org/record/4735821">GTDB Version
202: Genome Taxonomy Database</a></strong> (<a
href="http://gtdb.ecogenomic.org/">More info on GTDB</a>)</p>
<ul>
<li>Version 86 for <a
href="https://zenodo.org/record/2541239"><code>assignTaxonomy</code></a>
and <a
href="https://zenodo.org/record/2658728#.XM7qtBNKhYh"><code>assignSpecies</code></a></li>
</ul></li>
<li><p>RefSeq + RDP (NCBI RefSeq 16S rRNA database supplemented by
RDP)</p>
<ul>
<li><a href="https://zenodo.org/record/2541239">Reference files
formatted for <code>assignTaxonomy</code></a></li>
<li><a href="https://zenodo.org/record/2658728#.XM7qtBNKhYh">Reference
files formatted for <code>assignSpecies</code></a></li>
</ul></li>
<li><p><a href="http://doi.org/10.5281/zenodo.159205">HitDB version
1</a> (Human InTestinal 16S rRNA)</p></li>
<li><p><a href="https://doi.org/10.5281/zenodo.7380291">Human Oral
Microbiome Database: HOMD</a></p></li>
<li><p><a href="https://midasfieldguide.org/guide/downloads">MiDAS:
Field Guide to the Microbes of Activated Sludge and Anaerobic
Digesters</a></p></li>
<li><p><a href="http://www.reference-midori.info/index.html">MIDORI
Reference 2</a> (for taxonomic assignments of Eukaryota mitochondrial
DNA sequences)</p></li>
<li><p><a href="https://doi.org/10.5281/zenodo.835612">RDP fungi LSU
trainset 11</a></p></li>
<li><p><a href="https://zenodo.org/record/1447330">Silva Eukaryotic 18S,
v132 &amp; v128</a></p></li>
<li><p><a href="https://doi.org/10.5281/zenodo.3958364">nifH ARB,
version 1</a></p></li>
<li><p><a href="https://github.com/vaulot/pr2_database/releases">PR2
version 5.0.0+</a>. SEE NOTE BELOW.</p></li>
</ul>
<p><strong>Note:</strong> PR2 has different <code>taxLevels</code> than
the dada2 default. When assigning taxonomy against PR2, use the
following:
<code>assignTaxonomy(..., taxLevels = c("Domain","Supergroup","Division","Subdivision", "Class","Order","Family","Genus","Species"))</code></p>
<hr />
<p>Many thanks to the folks at RDP, Silva, GreenGenes, UNITE, GTDB, PR2
and others for making these amazing reference datbases available to the
community. We created the dada2-compatible training fastas from <a
href="https://www.arb-silva.de/no_cache/download/archive/release_138_1/Exports/">the
Silva NR99 and taxonomy files</a>, the <a
href="https://sourceforge.net/p/rdp-classifier/news/2020/07/rdp-classifier-213-july-2020-release-note/">RDP
trainset 18 database</a>, and <a
href="ftp://greengenes.microbio.me/greengenes_release/gg_13_8_otus/">the
GreenGenes 13.8 OTUs clustered at 97%</a>.</p>
<p>The scripts to parse and produce the officially maintained training
fastas are included as private functions within the dada2 R package.
Inspect <code>taxonomy.R</code> to see the code. Currently training
fastas for <code>assignSpecies</code> are only available for the Silva
and RDP databases.</p>
</div>
<div id="formatting-custom-databases" class="section level2">
<h2>Formatting custom databases</h2>
<p>Custom databases can be used with <code>assignTaxonomy</code> and
<code>assignSpecies</code> provided they are in the dada2-compatible
training fasta format. We thank the many contributors who have created
custom databases suitable for use with DADA2, several of which are
linked above.</p>
<p><code>assignTaxonomy(...)</code> expects a training fasta file (or
compressed fasta file) in which the taxonomy corresponding to each
sequence is encoded in the id line in the following fashion (the second
sequence is not assigned down to level 6):</p>
<pre><code>&gt;Level1;Level2;Level3;Level4;Level5;Level6;
ACCTAGAAAGTCGTAGATCGAAGTTGAAGCATCGCCCGATGATCGTCTGAAGCTGTAGCATGAGTCGATTTTCACATTCAGGGATACCATAGGATAC
&gt;Level1;Level2;Level3;Level4;Level5;
CGCTAGAAAGTCGTAGAAGGCTCGGAGGTTTGAAGCATCGCCCGATGGGATCTCGTTGCTGTAGCATGAGTACGGACATTCAGGGATCATAGGATAC</code></pre>
<p><code>assignSpecies(...)</code> expects the training data to be
provided in the form of a fasta file (or compressed fasta file), with
the id line formatted as follows:</p>
<pre><code>&gt;ID Genus species
ACCTAGAAAGTCGTAGATCGAAGTTGAAGCATCGCCCGATGATCGTCTGAAGCTGTAGCATGAGTCGATTTTCACATTCAGGGATACCATAGGATAC
&gt;ID Genus species
CGCTAGAAAGTCGTAGAAGGCTCGGAGGTTTGAAGCATCGCCCGATGGGATCTCGTTGCTGTAGCATGAGTACGGACATTCAGGGATCATAGGATAC</code></pre>
</div>

<div class="container" style="width: 100%;color:grey;text-align:right">
  <hr>
  Maintained by Benjamin Callahan (benjamin DOT j DOT callahan AT gmail DOT com)
  <br> Documentation License: <a href="https://creativecommons.org/licenses/by/4.0">CC-BY 4.0</a>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
