---
title: "Taxonomic reference data"
---

The `assignTaxonomy` and `assignSpecies` functions require appropriately formatted fasta files describing the set of taxonomically assigned sequences to use as a training dataset. We provide appropriately formatted versions of several popular taxonomic databases, and describe the dada2-specific format for those who wish to use a custom database.

## DADA2-formatted reference databases

We maintain reference fastas for the three most common 16S databases: Silva, RDP and GreenGenes. The dada2 package recognizes and parses the General Fasta releases of the UNITE project for ITS taxonomic assignment. Formatted versions of other databases can be "contributed" and will be made available through this page if referencable by doi (eg. deposited at Zenodo or Figshare). 

*Please note that the files provided here are just derivative reformattings of these taxonomic databases. If using these files for taxonomic assignment, the source database should also be cited.*

--------------------------

**Maintained:**

* **[Silva version 138.1 - UPDATED Mar 10, 2021](https://doi.org/10.5281/zenodo.4587954)**, [version 132](https://doi.org/10.5281/zenodo.1172782), [version 128](https://doi.org/10.5281/zenodo.801831),  [version 123](http://dx.doi.org/10.5281/zenodo.158958)
  - NOTE: As of Silva version 138, the official DADA2-formatted reference fastas are optimized for classification of Bacteria and Archaea, and are not suitable for classifying Eukaryotes.


* **[RDP trainset 18](https://doi.org/10.5281/zenodo.4310150)**, [RDP trainset 16](https://doi.org/10.5281/zenodo.801827), [RDP trainset 14](http://dx.doi.org/10.5281/zenodo.158955)

* **[UNITE (use the General Fasta releases, "All eukaryotes")](https://unite.ut.ee/repository.php)**

* *Deprecated: [GreenGenes version 13.8](http://dx.doi.org/10.5281/zenodo.158955) (the source GreenGenes database is no longer being maintained)*

**Contributed**:

* **[GTDB Version 202: Genome Taxonomy Database](https://zenodo.org/record/4735821)** ([More info on GTDB](http://gtdb.ecogenomic.org/)) 
    + Version 86 for [`assignTaxonomy`](https://zenodo.org/record/2541239) and [`assignSpecies`](https://zenodo.org/record/2658728#.XM7qtBNKhYh)

* RefSeq + RDP (NCBI RefSeq 16S rRNA database supplemented by RDP)
    + [Reference files formatted for `assignTaxonomy`](https://zenodo.org/record/2541239)
    + [Reference files formatted for `assignSpecies`](https://zenodo.org/record/2658728#.XM7qtBNKhYh)  
    
* [HitDB version 1](http://doi.org/10.5281/zenodo.159205) (Human InTestinal 16S rRNA)

* [Human Oral Microbiome Database: HOMD](https://doi.org/10.5281/zenodo.7380291)

* [MiDAS: Field Guide to the Microbes of Activated Sludge and Anaerobic Digesters](https://midasfieldguide.org/guide/downloads)

* [MIDORI Reference 2](http://www.reference-midori.info/index.html) (for taxonomic assignments of Eukaryota mitochondrial DNA sequences)

* [RDP fungi LSU trainset 11](https://doi.org/10.5281/zenodo.835612)

* [Silva Eukaryotic 18S, v132 & v128](https://zenodo.org/record/1447330)

* [nifH ARB, version 1](https://doi.org/10.5281/zenodo.3958364)

* [PR2 version 5.0.0+](https://github.com/vaulot/pr2_database/releases). SEE NOTE BELOW.

**Note:** PR2 has different `taxLevels` than the dada2 default. When assigning taxonomy against PR2, use the following: `assignTaxonomy(..., taxLevels = c("Domain","Supergroup","Division","Subdivision", "Class","Order","Family","Genus","Species"))`


--------------------------

Many thanks to the folks at RDP, Silva, GreenGenes, UNITE, GTDB, PR2 and others for making these amazing reference datbases available to the community. We created the dada2-compatible training fastas from [the Silva NR99 and taxonomy files](https://www.arb-silva.de/no_cache/download/archive/release_138_1/Exports/), the [RDP trainset 18 database](https://sourceforge.net/p/rdp-classifier/news/2020/07/rdp-classifier-213-july-2020-release-note/), and [the GreenGenes 13.8 OTUs clustered at 97\%](ftp://greengenes.microbio.me/greengenes_release/gg_13_8_otus/).

The scripts to parse and produce the officially maintained training fastas are included as private functions within the dada2 R package. Inspect `taxonomy.R` to see the code. Currently training fastas for `assignSpecies` are only available for the Silva and RDP databases. 

## Formatting custom databases

Custom databases can be used with `assignTaxonomy` and `assignSpecies` provided they are in the dada2-compatible training fasta format. We thank the many contributors who have created custom databases suitable for use with DADA2, several of which are linked above.

`assignTaxonomy(...)` expects a training fasta file (or compressed fasta file) in which the taxonomy corresponding to each sequence is encoded in the id line in the following fashion (the second sequence is not assigned down to level 6):

```
>Level1;Level2;Level3;Level4;Level5;Level6;
ACCTAGAAAGTCGTAGATCGAAGTTGAAGCATCGCCCGATGATCGTCTGAAGCTGTAGCATGAGTCGATTTTCACATTCAGGGATACCATAGGATAC
>Level1;Level2;Level3;Level4;Level5;
CGCTAGAAAGTCGTAGAAGGCTCGGAGGTTTGAAGCATCGCCCGATGGGATCTCGTTGCTGTAGCATGAGTACGGACATTCAGGGATCATAGGATAC
```

`assignSpecies(...)` expects the training data to be provided in the form of a fasta file (or compressed fasta file), with the id line formatted as follows:

```
>ID Genus species
ACCTAGAAAGTCGTAGATCGAAGTTGAAGCATCGCCCGATGATCGTCTGAAGCTGTAGCATGAGTCGATTTTCACATTCAGGGATACCATAGGATAC
>ID Genus species
CGCTAGAAAGTCGTAGAAGGCTCGGAGGTTTGAAGCATCGCCCGATGGGATCTCGTTGCTGTAGCATGAGTACGGACATTCAGGGATCATAGGATAC
```
