<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Ben Callahan, Paul McMurdie, Susan Holmes" />


<title>Installation and Testing</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Installation and Testing</h1>
<h4 class="author"><em>Ben Callahan, Paul McMurdie, Susan Holmes</em></h4>
</div>


<div id="installation-instructions" class="section level2">
<h2>Installation Instructions</h2>
<div id="install-external-dependencies" class="section level3">
<h3>0. Install External Dependencies</h3>
<p>The following instructions assume that you have installed R and have installed <a href="https://developer.apple.com/xcode/downloads/">Xcode</a> (Mac) or <a href="http://cran.r-project.org/bin/windows/Rtools/">Rtools</a> (Windows); as well as <a href="http://bioconductor.org/install/">Bioconductor</a>.</p>
<p>For Bioconductor, start a fresh R session, and enter the following.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;http://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="dt">suppressUpdates =</span> <span class="ot">FALSE</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;ShortRead&quot;</span>, <span class="dt">suppressUpdates =</span> <span class="ot">FALSE</span>)</code></pre>
</div>
<div id="download-and-unzip." class="section level3">
<h3>1. Download and unzip.</h3>
<p>Download the zipped package (attached) and unzip it.</p>
</div>
<div id="attempt-installation-from-within-r" class="section level3">
<h3>2. Attempt installation from within R</h3>
<p>Start a fresh R session, and enter the following.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;path/to/dadac&quot;</span>,
                 <span class="dt">repos =</span> <span class="ot">NULL</span>,
                 <span class="dt">type =</span> <span class="st">&quot;source&quot;</span>,
                 <span class="dt">dependencies =</span> <span class="kw">c</span>(<span class="st">&quot;Depends&quot;</span>, <span class="st">&quot;Suggests&quot;</span>,<span class="st">&quot;Imports&quot;</span>))</code></pre>
<p>For example, if the dadac source code directory was in <code>~/github/dadac</code>, then the following would work.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;~/github/dadac&quot;</span>,
                 <span class="dt">repos =</span> <span class="ot">NULL</span>,
                 <span class="dt">type =</span> <span class="st">&quot;source&quot;</span>,
                 <span class="dt">dependencies =</span> <span class="kw">c</span>(<span class="st">&quot;Depends&quot;</span>, <span class="st">&quot;Suggests&quot;</span>,<span class="st">&quot;Imports&quot;</span>))</code></pre>
<p>Check Package Version</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">packageVersion</span>(<span class="st">&quot;dadac&quot;</span>)</code></pre>
<pre><code>## [1] '0.4.0'</code></pre>
</div>
<div id="troubleshoot-dependencies-optional" class="section level3">
<h3>3. Troubleshoot Dependencies (Optional)</h3>
<p>At this point, there may be complaints about missing dependencies. To install missing dependencies on either <a href="http://cran.r-project.org/">CRAN</a> or <a href="http://bioconductor.org/install/">Bioconductor</a>, start a fresh R session, and enter the following.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;http://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;missing_package_1&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;missing_package_2&quot;</span>)
<span class="co"># ... and so on</span></code></pre>
</div>
<div id="re-attempt-dadac-installation" class="section level3">
<h3>4. Re-attempt dadac Installation</h3>
<p>Now once again, after dependencies have been installed.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;path/to/dadac&quot;</span>,
                 <span class="dt">repos =</span> <span class="ot">NULL</span>,
                 <span class="dt">type =</span> <span class="st">&quot;source&quot;</span>,
                 <span class="dt">dependencies =</span> <span class="kw">c</span>(<span class="st">&quot;Depends&quot;</span>, <span class="st">&quot;Suggests&quot;</span>,<span class="st">&quot;Imports&quot;</span>))</code></pre>
<p>You should now be done, if installation was successful. If not, do panic, and post a message to the issue tracker.</p>
<hr />
</div>
</div>
<div id="testing-instructions" class="section level2">
<h2>Testing Instructions</h2>
<div id="suggestions" class="section level3">
<h3>0. Suggestions</h3>
<p>If youâ€™re here, the package installed which is already great! Any suggestions or issues with the following I would love to hear about.</p>
</div>
<div id="load-package-explore-function-documentation" class="section level3">
<h3>1. Load Package, Explore Function Documentation</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;dadac&quot;</span>)</code></pre>
<p>Now that the package is loaded, look at the help files for the main implemented functions.</p>
<pre class="sourceCode r"><code class="sourceCode r">?dada
?derepFastq
?importUniques</code></pre>
</div>
<div id="test-out-dada" class="section level3">
<h3>2. Test out DADA:</h3>
<p>THIS IS OLD!</p>
<pre class="sourceCode r"><code class="sourceCode r">score &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">5</span>, -<span class="dv">4</span>, -<span class="dv">4</span>, -<span class="dv">4</span>, -<span class="dv">4</span>, <span class="dv">5</span>, -<span class="dv">4</span>, 
                  -<span class="dv">4</span>, -<span class="dv">4</span>, -<span class="dv">4</span>, <span class="dv">5</span>, -<span class="dv">4</span>, -<span class="dv">4</span>, -<span class="dv">4</span>, -<span class="dv">4</span>, <span class="dv">5</span>),
                <span class="dt">nrow=</span><span class="dv">4</span>, <span class="dt">byrow=</span>T)
gap_penalty &lt;-<span class="st"> </span>-<span class="dv">8</span>
err_init &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="fl">0.991</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, 
                     <span class="fl">0.003</span>, <span class="fl">0.991</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, 
                     <span class="fl">0.003</span>, <span class="fl">0.991</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, 
                     <span class="fl">0.003</span>, <span class="fl">0.991</span>),
                   <span class="dt">nrow=</span><span class="dv">4</span>, <span class="dt">byrow=</span>T)
test_fastq_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;test-nonunique.fastq.gz&quot;</span>, <span class="dt">package=</span><span class="st">&quot;dadac&quot;</span>)
uniques &lt;-<span class="st"> </span><span class="kw">derepFastq</span>(<span class="dt">fl =</span> test_fastq_file)
uniques.dada &lt;-<span class="st"> </span><span class="kw">dada</span>(<span class="dt">uniques =</span> uniques$uniques, <span class="dt">quals =</span> uniques$quals)
uniques.dada.consistent &lt;-<span class="st"> </span><span class="kw">dada</span>(uniques, err_init, score, gap_penalty, <span class="dt">self_consist =</span> <span class="ot">TRUE</span>)
<span class="kw">length</span>(uniques)
<span class="kw">length</span>(uniques.dada$genotypes)
<span class="kw">length</span>(uniques.dada.consistent$genotypes)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Out of `r length(uniques)` original unique sequences,</span>
<span class="co"># did you get `r length(uniques.dada$genotypes)`</span>
<span class="co"># genotypes in `uniques.dada$genotypes` and</span>
<span class="co"># `r length(uniques.dada.consistent$genotypes)` in `uniques.dada.consistent`?</span></code></pre>
</div>
<div id="performance-check-dada" class="section level3">
<h3>3. Performance check DADA:</h3>
<p>THIS IS OLD!!</p>
<pre class="sourceCode r"><code class="sourceCode r">group_0_25_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;group_0-25.uniques.gz&quot;</span>, <span class="dt">package=</span><span class="st">&quot;dadac&quot;</span>)
bu =<span class="st"> </span><span class="kw">importUniques</span>(group_0_25_file)
<span class="kw">length</span>(bu)
<span class="kw">library</span>(<span class="st">&quot;microbenchmark&quot;</span>)
<span class="kw">microbenchmark</span>(<span class="kw">dada</span>(bu, <span class="dt">err =</span> err_init, ), <span class="dt">times=</span><span class="dv">10</span>)</code></pre>
<p>How long does this expression take? (Hopefully ~10 seconds or less per evaluation.)</p>
<p>THIS IS OLD!</p>
<pre class="sourceCode r"><code class="sourceCode r">group_0_30_file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;group_0-30.uniques.gz&quot;</span>, <span class="dt">package=</span><span class="st">&quot;dadac&quot;</span>)
bu =<span class="st"> </span><span class="ot">NULL</span>
bu &lt;-<span class="st"> </span><span class="kw">importUniques</span>(group_0_30_file)
<span class="kw">length</span>(bu)
<span class="kw">microbenchmark</span>(<span class="kw">dada</span>(bu, err_init, score, gap_penalty), <span class="dt">times=</span><span class="dv">10</span>)</code></pre>
<p>And this?</p>
</div>
<div id="performance-check" class="section level3">
<h3>4. Performance Check</h3>
<p>THIS IS OLD!</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;ggplot2&quot;</span>)
<span class="kw">length</span>(bu)
<span class="co"># The max number of sequences from `bu` to include in this example</span>
N =<span class="st"> </span>1000L
<span class="co"># Set random see for repoducibility of example</span>
<span class="kw">set.seed</span>(711L)
<span class="co"># Randomly subsample from `bu`, because `bu` </span>
<span class="co"># on its own is large and takes a long time</span>
buSub0 =<span class="st"> </span>bu[<span class="kw">sample</span>(<span class="dt">x =</span> <span class="kw">length</span>(bu), <span class="dt">size =</span> N, <span class="dt">replace =</span> <span class="ot">FALSE</span>)]
<span class="co"># Run and report time elapsed</span>
<span class="kw">system.time</span>({
  df &lt;-<span class="st"> </span><span class="kw">calibrate_kmers</span>(<span class="dt">seqs =</span> <span class="kw">names</span>(buSub0),
                        <span class="dt">score =</span> <span class="kw">get_dada_opt</span>(<span class="st">&quot;SCORE_MATRIX&quot;</span>),
                        <span class="dt">gap =</span>  -<span class="dv">8</span>,
                        <span class="dt">max_aligns =</span> <span class="dv">100000</span>)
})
<span class="co"># Make a summary plot</span>
<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x=</span>kmer, <span class="dt">y=</span>align)) +<span class="st"> </span><span class="kw">geom_bin2d</span>()
<span class="co"># Som alternative plots</span>
<span class="co"># ggplot(df, aes(x=kmer, y=align)) + geom_hex()</span>
<span class="co"># ggplot(df, aes(x=kmer, y=align)) + geom_density2d()</span></code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
